### `Llama2Math`

Allowing an [agent](example.ipynb) to solve mathematical problems can be done using a specialized LLM. This LLM is used to generate Python code which is subsequently evaluated using a Python interpreter to obtain the result of the generated calculation. This behaviour is implemented in [Llama2Math](https://github.com/krasserm/grammar-based-agents/blob/master/gba/math.py) which uses an instruction tuned LLM to interpret and evaluate mathematical queries. The [Llama2Math](https://github.com/krasserm/grammar-based-agents/blob/master/gba/math.py) interface accepts arbitrary mathematical problems (e.g. `What is the result of 2 times 5, then raised to the power of 0.8 and divided by 2`), translates the problem into runnable Python code using the configured LLM and executes the code to obtain the result.

Currently, only CodeLlama models (e.g. [CodeLlama-7b model](https://huggingface.co/TheBloke/CodeLlama-7B-Instruct-GGUF)) are supported.

The following examples use a 4-bit quantized [CodeLlama-7b model](https://huggingface.co/TheBloke/CodeLlama-7B-Instruct-GGUF) (launch instructions [here](https://github.com/krasserm/grammar-based-agents/blob/master/README.md#getting-started)).

In [1]:
from gba.llm import LlamaCppClient  
from gba.math import Llama2Math

math_llm = LlamaCppClient(url="http://localhost:8088/completion", temperature=-1)

math_model = Llama2Math(llm=math_llm)

In [2]:
math_model.run("What is the result of 2 times 5, then raised to the power of 0.8 and divided by 2")

Executing Llama2Math Python code:
```
result = (2 * 5) ** 0.8 / 2
```


'3.15479'

This example shows the generated Python code to solve the mathematical problem and the result of the code execution which is evaluated to `3.15479`.

The following examples demonstrate other types of mathematical queries:

In [3]:
math_model.run("What is 73 raised to the 0.24 power?")

Executing Llama2Math Python code:
```
result = pow(73, 0.24)
```


'2.80025'

In [4]:
math_model.run("27^0.24")

Executing Llama2Math Python code:
```
result = pow(27, 0.24)
```


'2.20560'

In [5]:
math_model.run("Give me the result of 3*4")

Executing Llama2Math Python code:
```
result = 3 * 4
```


'12.00000'

In [6]:
math_model.run("What is (2+17+3) raised to the power of 0.5?")

Executing Llama2Math Python code:
```
result = (2+17+3) ** 0.5
```


'4.69042'

In [7]:
math_model.run("What is the result of three plus four, divided by two, then multiplied by seven point five-seven and then raised to the power of 1.34")

Executing Llama2Math Python code:
```
result = (3 + 4) / 2 * 7.57 ** 1.34
```


'52.72985'